### DISCUSSION TITLE
Executing Custom API Calls based on Query Response

### DISCUSSION THREAD

**Discussion by SiPurdy** [MarkMpn/Sql4Cds#468] [Q&A]
I have a CustomAPI call which has the following inputs...
- Config: String
- DisablePlugins: boolean
- Target: EntityReference (expando)
    The idea is that the CustomAPI can be invoked on any record.

... and no outputs.

The call can be executed using the following syntax without issue.
```exec <customapiname> ....```

The question is whether it's possible to run the individually customapi for each record returned from a query?

I have tried Cross Apply to no luck and Sql4Cds doesn't do [Cursors](https://learn.microsoft.com/en-us/sql/t-sql/language-elements/declare-cursor-transact-sql?view=sql-server-ver16).

**Comment by MarkMpn** [comment]
Cursors are the real solution here, and one I’d like to implement in the future. For now, #424 shows one method of looping through data and performing an action. It’s not as efficient as a cursor would be but should be good enough for small data sets.

**Reply by SiPurdy** [reply]
Is it possible that a signature change of the customapi call could help? For example having it output say a boolean for whether the day of the week ends in a Y, and use it as a function for a select column, something like:

``` sql
select
    contactid
    , fullname
    , <customapiname>(@Target = contactid, ...) as Processed
from contact
where
    statecode=0 and
    firstname in ('Alice', 'Bob')
```

Irrespective I can see use in adding of scalar customapi calls like this to column outputs, especially if they can be referenced in where clauses (yes it would mean running the customapi for all records but this could be optimised by taking the where clauses which can be delegated, or those processed clientside that aren't dependent on the customapi). This _might?_ be easier to implement than cursors and would be a quicker win?

**Reply by MarkMpn** [reply]
Yes that would work. If changing the API signature is possible, adding an output parameter would let SQL 4 CDS call it as table-valued function which you could then reference with `CROSS APPLY` and use immediately without any changes to SQL 4 CDS.

**Reply by SiPurdy** [reply]
That works great-ish.

One concern, what controls the order of the parameters as it seems they go in comma separated and I'm worried that the order might differ between environments. Are you fixing the order or is it possible to specify the name of parameters going in, eg:
```... CROSS APPLY (SELECT * FROM myProc(@Trees = 1, @Flowers = 2, @Birds = 3)) ...```
vs
```... CROSS APPLY (SELECT * FROM myProc(1, 2, 3)) ...```

**Reply by MarkMpn** [reply]
Naming parameters in a call to a table-valued function is not supported in TSQL. The order of the parameters is defined by the `position` attribute of the `sdkmessagerequestfield`. I don't believe this is exposed by the UI for creating Custom API request parameters, but should be consistent for APIs moved as part of solutions.

**Reply by SiPurdy** [reply]
> but should be consistent for APIs moved as part of solutions

There's no specific positional information in the solution exports having poked around the customapi.xml or customapirequestparameter.xml files which does leave me a little concerned that the order might be a bit more fuzzy, eg alphabetical based on the order that the parameters are first created meaning an updated deployment with a new parameter might have a different order compared to a fresh install of the same version.

I've just been burnt so many times that if something isn't specified within the solution exports you can't rely on it.

I'll see if experience mirrors expectations but there may be situations when diverging from existing supported TSQL syntax might be required to make it possible to create scripts that don't need changing per environment (eg parameter ordering).

**Reply by SiPurdy** [reply]
Not good news. Took my managed solution with the customapi and imported into a clean environment, sdkmessagerequestfield records are as follows:

| parameter | position |
|---|---|
| DeleteAudit | 1 |
| Config | 2 |
| DisablePluginWorkflowFlow | 3 |
| Target | 4 |

Deleted the managed solution and re-imported the same solution file:

| parameter | position |
|---|---|
| DeleteAudit | 1 |
| DisablePluginWorkflowFlow | 2 |
| Config | 3 |
| Target | 4 |

A third run ended up with a different order. 

In short, there is zero guarantee that the position on sdkmessagerequestfield is going to match between environments which means care needs to be taken when running cross-apply scripts across environments.

**Reply by MarkMpn** [reply]
> there may be situations when diverging from existing supported TSQL syntax might be required to make it possible to create scripts that don't need changing per environment (eg parameter ordering).

SQL 4 CDS uses the ScriptDOM library to parse and manipulate queries. This is built specifically for the TSQL syntax. Any deviation from that syntax would require completely replacing that library with a custom parser, probably created as a fork of the main ScriptDOM code. I'd be happy to review a PR that implements such a parser, but it's not something I've got time to work on myself.

**Reply by SiPurdy** [reply]
Don't blame you on that one. 20-odd years since I played with YACC, and 5 years since I tried some stuff using ANTLR.

Is it viable to ignore the position column on sdkmessagerequestfield and process the parameters alphabetically as that will be deterministic between deployments?
